LSPS2/README.md

```diff
@@ -0,0 +1,736 @@
+    `promise` before.
+    It does not have to be third-party-verifiable (i.e. it is OK
+    if only the LSP can do this verification using private information
+    only the LSP knows).
+  * Unforgeable - only the LSP can generate the `promise`.
+    Knowledge of the `promise` for one set of parameters should not
+    allow anyone else to generate the `promise` for a different set
+    of parameters.
+  * Committing - the `promise` commits to a specific set of parameters;
+    if any of the parameters is changed, the `promise` validation would
+    fail.
+* MUST ensure the promise is no longer than 512 bytes (not including the
+  `"` delimiters) when encoded as a UTF8-encoded JSON string.
+* SHOULD avoid characters outside of the printable ASCII range.
+* SHOULD avoid characters that would require escapes in JSON format.
+
+> **Rationale** The method to generate `promise` is left vague in order
+> to allow LSP implementors the flexibility to structure their code
+> as necessary.
+> The above specification suggests the use of any MAC, which would cover
+> the above requirements, and requires a symmetric key known only by
+> the LSP.
+> Alternately, an LSP might be componentized and use private-public
+> key cryptography, with a signature as the `promise`, so that the
+> component that generates `opening_fee_params` objects is the only
+> one with access to a specific private key, while the rest of the LSP,
+> including the `lsps2.buy` component, only knows the public key.
+> Another alternative for a componentized LSP would be to have the
+> `opening_fee_params` generator know the public key of the
+> `opening_fee_params` validator; the generator creates an ephemeral
+> keypair and uses the ephemeral privkey with the validator pubkey
+> to create a session key for a MAC, then concatenate the ephemeral
+> pubkey with the MAC as the `promise`; the validator takes its
+> private key and uses it with the ephemeral pubkey to regenerate
+> the same session key and validates the MAC.
+
+Clients, when reading the `promise` feild:
+
+* MUST abort this flow if `promise` exceeds 512 bytes (not including the
+  `"` delimiters) when encoded as a UTF8-encoded JSON string.
+* MUST NOT parse or interpret this string, and only provide it
+  verbatim later.
+
+> **Rationale** By using a promise that must be remembered by the
+> client, the LSP does not have to remember what it offered via this
+> API, reducing state storage, especially if the client then decides
+> that the resulting `opening_fee` is too high and ultimately decides
+> not to continue this flow.
+> By using a cryptographically-verifiable method, the LSP can ensure
+> that it will only honour `opening_fee_params` it actually offered.
+> A size limit is imposed so that LSPs cannot burden clients with
+> unreasonable storage requirements.
+
+LSPs, and future versions of this specfication, MUST NOT add fields
+to the `opening_fee_params` object other than the fields listed
+above.
```

Comment:
If the client makes sure that it sends the exact `opening_fee_params` object as was returned in the `getinfo` response, it may be possible to add fields? It would be nice, if we're using json, to allow json dynamics as expected. Which is additional fields are always allowed in json, you just can't remove any.

Reply:
For this particular case, the `promise` commits to a specific set of parameters.  Adding new fields is thus of dubious use, as the `promise` would not commit to them and the LSP cannot be certain if it was generated by the LSP or malleated by the client.

Reply:
The client also kinda needs to really understand what the feerates are, if any new parameters are added which change the feerate computation, that changes how the client can interpret this.  Which is why a later rationale section recommends just adding a new `opening_fee_params_menu_v2` if necessary, so that the client can ignore that entire array if it does not know the v2 feerate computation.

Reply:
Yeah I agree it becomes too hard if you want to support multiple versions with a single object there. Even if the promise would commit to all fields in the `opening_fee_params` it's probably too hard to support multiple versions that way. 

If the subversions, so the versions of a specific LSPS, would be returned in the `lsps0.listprotocols` call, the `getinfo` call here could be requested with a specific version. That could give us an easier upgrade path. I can imagine changing the opening_fee_params would be one of the changes made to this spec.

Reply:
Maybe just separate the `getversions` call here then?

Reply:
Added `getversions` call.

LSP is still not allowed to add additional fields in `opening_fee_params`, but at least future LSPS versions are allowed to add additional fields.

Reply:
Done.